cpp/src/arrow/gpu/cuda_context.cc

```diff
@@ -293,11 +334,32 @@ std::shared_ptr<CudaDevice> CudaMemoryManager::cuda_device() const {
   return checked_pointer_cast<CudaDevice>(device_);
 }
 
+Result<std::shared_ptr<Device::SyncEvent>> CudaMemoryManager::MakeDeviceSyncEvent() {
+  ARROW_ASSIGN_OR_RAISE(auto context, cuda_device()->GetContext());
+  ContextSaver set_temporary((CUcontext)(context.get()->handle()));
+
+  // TODO: event creation flags??
```

Comment:
This one matters a bit more since the CUDA event creation flags can be used to control behavior as well as can have non-negligible performance differences (disabling timing can yield perf).

That being said, I'd say we should again punt until there's proof of its need as someone can always use the underlying types directly as needed.